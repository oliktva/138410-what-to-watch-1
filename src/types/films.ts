import PropTypes from 'prop-types';

export const filmPropTypes = PropTypes.shape({
  id: PropTypes.number.isRequired,
  name: PropTypes.string.isRequired,
  img: PropTypes.string.isRequired,
  video: PropTypes.string.isRequired
});

export const filmsPropTypes = PropTypes.arrayOf(filmPropTypes.isRequired);

export interface FilmProps {
  name: string;
  img: string;
  video: string;
  id: number;
}
