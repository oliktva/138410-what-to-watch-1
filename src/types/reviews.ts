export interface ReviewProps {
  comment: string;
  date: string;
  id: number;
  rating: number;
  user: {
    id: number;
    name: string;
  };
}
