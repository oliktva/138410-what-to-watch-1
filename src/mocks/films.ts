import {FilmProps} from "src/types/films";

export const films: FilmProps[] = [{
  id: 1,
  name: `Fantastic Beasts`,
  img: `img/fantastic-beasts-the-crimes-of-grindelwald.jpg`,
  video: `https://download.blender.org/durian/trailer/sintel_trailer-480p.mp4`
}, {
  id: 2,
  name: `Bohemian Rhapsody`,
  img: `img/bohemian-rhapsody.jpg`,
  video: `https://download.blender.org/durian/trailer/sintel_trailer-480p.mp4`
}, {
  id: 3,
  name: `Macbeth`,
  img: `img/macbeth.jpg`,
  video: `https://download.blender.org/durian/trailer/sintel_trailer-480p.mp4`
}, {
  id: 4,
  name: `Aviator`,
  img: `img/aviator.jpg`,
  video: `https://download.blender.org/durian/trailer/sintel_trailer-480p.mp4`
}, {
  id: 5,
  name: `We need to talk about Kevin`,
  img: `img/we-need-to-talk-about-kevin.jpg`,
  video: `https://download.blender.org/durian/trailer/sintel_trailer-480p.mp4`
}, {
  id: 6,
  name: `What We Do in the Shadows`,
  img: `img/what-we-do-in-the-shadows.jpg`,
  video: `https://download.blender.org/durian/trailer/sintel_trailer-480p.mp4`
}, {
  id: 7,
  name: `Revenant`,
  img: `img/revenant.jpg`,
  video: `https://download.blender.org/durian/trailer/sintel_trailer-480p.mp4`
}, {
  id: 8,
  name: `Johnny English`,
  img: `img/johnny-english.jpg`,
  video: `https://download.blender.org/durian/trailer/sintel_trailer-480p.mp4`
}, {
  id: 9,
  name: `Shutter Island`,
  img: `img/shutter-island.jpg`,
  video: `https://download.blender.org/durian/trailer/sintel_trailer-480p.mp4`
}, {
  id: 10,
  name: `Pulp Fiction`,
  img: `img/pulp-fiction.jpg`,
  video: `https://upload.wikimedia.org/wikipedia/commons/transcoded/b/b3/Big_Buck_Bunny_Trailer_400p.ogv/Big_Buck_Bunny_Trailer_400p.ogv.360p.webm`
}, {
  id: 11,
  name: `No Country for Old Men`,
  img: `img/no-country-for-old-men.jpg`,
  video: `https://upload.wikimedia.org/wikipedia/commons/transcoded/b/b3/Big_Buck_Bunny_Trailer_400p.ogv/Big_Buck_Bunny_Trailer_400p.ogv.360p.webm`
}, {
  id: 12,
  name: `Snatch`,
  img: `img/snatch.jpg`,
  video: `https://upload.wikimedia.org/wikipedia/commons/transcoded/b/b3/Big_Buck_Bunny_Trailer_400p.ogv/Big_Buck_Bunny_Trailer_400p.ogv.360p.webm`
}, {
  id: 13,
  name: `Moonrise Kingdom`,
  img: `img/moonrise-kingdom.jpg`,
  video: `https://upload.wikimedia.org/wikipedia/commons/transcoded/b/b3/Big_Buck_Bunny_Trailer_400p.ogv/Big_Buck_Bunny_Trailer_400p.ogv.360p.webm`
}, {
  id: 14,
  name: `Seven Years in Tibet`,
  img: `img/seven-years-in-tibet.jpg`,
  video: `https://upload.wikimedia.org/wikipedia/commons/transcoded/b/b3/Big_Buck_Bunny_Trailer_400p.ogv/Big_Buck_Bunny_Trailer_400p.ogv.360p.webm`
}, {
  id: 15,
  name: `Midnight Special`,
  img: `img/midnight-special.jpg`,
  video: `https://upload.wikimedia.org/wikipedia/commons/transcoded/b/b3/Big_Buck_Bunny_Trailer_400p.ogv/Big_Buck_Bunny_Trailer_400p.ogv.360p.webm`
}, {
  id: 16,
  name: `War of the Worlds`,
  img: `img/war-of-the-worlds.jpg`,
  video: `https://upload.wikimedia.org/wikipedia/commons/transcoded/b/b3/Big_Buck_Bunny_Trailer_400p.ogv/Big_Buck_Bunny_Trailer_400p.ogv.360p.webm`
}, {
  id: 17,
  name: `Dardjeeling Limited`,
  img: `img/dardjeeling-limited.jpg`,
  video: `https://upload.wikimedia.org/wikipedia/commons/transcoded/b/b3/Big_Buck_Bunny_Trailer_400p.ogv/Big_Buck_Bunny_Trailer_400p.ogv.360p.webm`
}, {
  id: 18,
  name: `Orlando`,
  img: `img/orlando.jpg`,
  video: `https://upload.wikimedia.org/wikipedia/commons/transcoded/b/b3/Big_Buck_Bunny_Trailer_400p.ogv/Big_Buck_Bunny_Trailer_400p.ogv.360p.webm`
}, {
  id: 19,
  name: `Mindhunter`,
  img: `img/mindhunter.jpg`,
  video: `https://upload.wikimedia.org/wikipedia/commons/transcoded/b/b3/Big_Buck_Bunny_Trailer_400p.ogv/Big_Buck_Bunny_Trailer_400p.ogv.360p.webm`
}, {
  id: 20,
  name: `Midnight Special`,
  img: `img/midnight-special.jpg`,
  video: `https://upload.wikimedia.org/wikipedia/commons/transcoded/b/b3/Big_Buck_Bunny_Trailer_400p.ogv/Big_Buck_Bunny_Trailer_400p.ogv.360p.webm`
}];
