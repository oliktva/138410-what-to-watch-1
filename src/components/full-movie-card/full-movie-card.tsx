import React, {Fragment, PureComponent, ReactElement} from 'react';
import {connect, MapDispatchToProps} from 'react-redux';
import {compose} from 'redux';
import PropTypes from 'prop-types';

import {Operation} from 'src/reducers/films/films';
import {formatTime, formatLocaleTime} from 'src/helpers/time-helper';
import {getRatingDescription} from 'src/helpers/rating-helper';
import withActiveTab from 'src/hocs/with-active-tab/with-active-tab';

import needRenderPlayer from 'src/hocs/need-render-player/need-render-player';
import Player from 'src/components/player/player';
import BackgroundImage from 'src/components/background-image/background-image';
import MoviePoster from 'src/components/movie-poster/movie-poster';
import CardButtons from 'src/components/card-buttons/card-buttons';
import MovieMeta from 'src/components/movie-meta/movie-meta';
import SmallFullMovieCardsList from 'src/components/small-movie-cards-list/small-movie-cards-list';
import Footer from 'src/components/footer/footer';

import {State, ThunkDispatch} from 'src/types/reducer';
import {FilmProps, filmPropTypes, filmsPropTypes, ReviewProps, reviewsPropTypes} from 'src/types/films';

const TABS = ['overview', 'details', 'reviews'];

interface DispatchProps {
  addToFavorites: (filmId: number) => Promise<void>;
  removeFromFavorites: (filmId: number) => Promise<void>;
}

interface WithActiveTabProps {
  activeTab: string;
  setActiveTab: (tab: string) => void;
}

interface NeedRenderPlayerProps {
  withPlayer: boolean;
  toggleRenderPlayer: () => void;
}

interface PartProps {
  header: ReactElement;
  film: FilmProps | null;
  reviews: ReviewProps[];
  relatedFilms: FilmProps[];
}

type OwnProps = PartProps & WithActiveTabProps & NeedRenderPlayerProps;

type Props = OwnProps & DispatchProps;

const propTypes = {
  header: PropTypes.element.isRequired,
  film: filmPropTypes.isRequired,
  relatedFilms: filmsPropTypes.isRequired,
  reviews: reviewsPropTypes.isRequired,
  withPlayer: PropTypes.bool.isRequired,
  toggleRenderPlayer: PropTypes.func.isRequired,
  addToFavorites: PropTypes.func.isRequired,
  removeFromFavorites: PropTypes.func.isRequired,
  isFull: PropTypes.bool,
  tabs: PropTypes.arrayOf(PropTypes.string),
  activeTab: PropTypes.string,
  setActiveTab: PropTypes.func
};

const defaultProps = {
  isFull: false
}

class FullMovieCard extends PureComponent<Props> {
  public static propTypes = propTypes;
  public static defaultProps = defaultProps;

  public constructor(props: Props) {
    super(props);

    this._handlePlay = this._handlePlay.bind(this);
    this._handleToggleFavorite = this._handleToggleFavorite.bind(this);
    this._handleTabClick = this._handleTabClick.bind(this);
  }

  public render(): ReactElement | null {
    const {
      film,
      reviews,
      withPlayer,
      toggleRenderPlayer,
      header,
      activeTab,
      relatedFilms
    } = this.props;

    if (!film) {
      return null;
    }

    return (
      <Fragment>
        {withPlayer && <Player film={film} closePlayer={toggleRenderPlayer} />}
        <section className="movie-card movie-card--full" style={{backgroundColor: film.backgroundColor}}>
          <div className="movie-card__hero">
            <BackgroundImage image={film.backgroundImage} name={film.name} />
            <h1 className="visually-hidden">WTW</h1>
            {header}
            <div className="movie-card__wrap">
              <div className="movie-card__desc">
                <MovieMeta film={film} />
                <CardButtons
                  film={film}
                  onPlayButtonClick={this._handlePlay}
                  onFavoritesToggle={this._handleToggleFavorite}
                  isFull
                />
              </div>
            </div>
          </div>
          <div className="movie-card__wrap movie-card__translate-top">
            <div className="movie-card__info">
              <MoviePoster image={film.posterImage} name={film.name} kind="big" />
              <div className="movie-card__desc">
                {this._renderNav()}
                {activeTab === `overview` && this._renderOverview(film)}
                {activeTab === `details` && this._renderDetails(film)}
                {activeTab === `reviews` && this._renderReviews(reviews)}
              </div>
            </div>
          </div>
        </section>
        {this._renderMoreLikeThis(relatedFilms)}
      </Fragment>
    );
  }

  private _handlePlay(): void {
    this.props.toggleRenderPlayer();
  }

  private _handleToggleFavorite(): void {
    const {film, addToFavorites, removeFromFavorites} = this.props;

    if (film) {
      if (film.isFavorite) {
        removeFromFavorites(film.id);
      } else {
        addToFavorites(film.id);
      }
    }
  }

  private _handleTabClick(tab: string): () => void {
    const {setActiveTab} = this.props;

    return () => {
      setActiveTab(tab);
    }
  }

  private _renderStarring(starring: string[], {isLine}: {isLine: boolean} = {isLine: false}): ReactElement[] {
    return starring.map((s: string, index: number): ReactElement => {
      if (index === starring.length - 1) {
        return isLine ? <Fragment key={s}>{s} and other</Fragment> : <Fragment key={s}>{s}</Fragment>;
      } else {
        return isLine ? <Fragment key={s}>{s}, </Fragment> : <Fragment key={s}>{s}, <br/></Fragment>;
      }
    });
  }

  private _renderNav(): ReactElement {
    const {activeTab} = this.props;

    return (
      <nav className="movie-nav movie-card__nav">
        <ul className="movie-nav__list">
          {TABS.map(tab => (
            <li key={tab} className={`movie-nav__item ${tab === activeTab ? 'movie-nav__item--active' : ''}`}>
              <span className="movie-nav__link" onClick={this._handleTabClick(tab)}>{tab}</span>
            </li>
          ))}
        </ul>
      </nav>
    );
  }
  
  private _renderOverview(film: FilmProps): ReactElement {
    return (
      <Fragment>
        <div className="movie-rating">
          <div className="movie-rating__score">{film.rating}</div>
          <p className="movie-rating__meta">
            <span className="movie-rating__level">{getRatingDescription(film.rating)}</span>
            <span className="movie-rating__count">{film.scoresCount.toLocaleString()} ratings</span>
          </p>
        </div>

        <div className="movie-card__text">
          <p>{film.description}</p>

          <p className="movie-card__director"><strong>Director: {film.director}</strong></p>

          <p className="movie-card__starring"><strong>Starring: {this._renderStarring(film.starring, {isLine: true})}</strong></p>
        </div>
      </Fragment>
    );
  }

  private _renderDetails(film: FilmProps): ReactElement {
    return (
      <div className="movie-card__text movie-card__row">
        <div className="movie-card__text-col">
          <p className="movie-card__details-item">
            <strong className="movie-card__details-name">Director</strong>
            <span className="movie-card__details-value">{film.director}</span>
          </p>
          <p className="movie-card__details-item">
            <strong className="movie-card__details-name">Starring</strong>
            <span className="movie-card__details-value">
              {this._renderStarring(film.starring)}
            </span>
          </p>
        </div>
        <div className="movie-card__text-col">
          <p className="movie-card__details-item">
            <strong className="movie-card__details-name">Run Time</strong>
            <span className="movie-card__details-value">
              {formatTime(film.runTime * 60, {withSeconds: false})}
            </span>
          </p>
          <p className="movie-card__details-item">
            <strong className="movie-card__details-name">Genre</strong>
            <span className="movie-card__details-value">{film.genre}</span>
          </p>
          <p className="movie-card__details-item">
            <strong className="movie-card__details-name">Released</strong>
            <span className="movie-card__details-value">{film.released}</span>
          </p>
        </div>
      </div>
    );
  }

  private _renderReview(review: ReviewProps): ReactElement {
    return (
      <div className="review" key={review.id}>
        <blockquote className="review__quote">
          <p className="review__text">{review.comment}</p>
          <footer className="review__details">
            <cite className="review__author">{review.user.name}</cite>
            <time className="review__date" dateTime="2016-12-24">{formatLocaleTime(review.date)}</time>
          </footer>
        </blockquote>
        <div className="review__rating">{review.rating}</div>
      </div>
    );
  }

  private _renderReviews(reviews: ReviewProps[]): ReactElement {
    const firstColumnsReviews = reviews.filter((_, index: number): boolean => (index)%2 === 0);
    const secondColumnsReviews = reviews.filter((_, index: number): boolean => (index)%2 === 1);

    return (
      <div className="movie-card__reviews movie-card__row">
        <div className="movie-card__reviews-col">
          {firstColumnsReviews.map((review: ReviewProps): ReactElement => this._renderReview(review))}
        </div>
        <div className="movie-card__reviews-col">
          {secondColumnsReviews.map((review: ReviewProps): ReactElement => this._renderReview(review))}
        </div>
      </div>
    );
  }

  private _renderMoreLikeThis(relatedFilms: FilmProps[]) {
    return (
      <div className="page-content">
        <section className="catalog catalog--like-this">
          <h2 className="catalog__title">More like this</h2>
          <SmallFullMovieCardsList films={relatedFilms} />
        </section>
        <Footer />
      </div>
    );
  }

}

export {FullMovieCard};

const mapDispatchToProps: MapDispatchToProps<DispatchProps, OwnProps> = (dispatch: ThunkDispatch): DispatchProps => ({
  addToFavorites: (filmId: number): Promise<void> => dispatch(Operation.addToFavorites(filmId)),
  removeFromFavorites: (filmId: number): Promise<void> => dispatch(Operation.addToFavorites(filmId))
});

const ConnectedComponent: any =
  compose(
    connect<{}, DispatchProps, OwnProps, State>(null, mapDispatchToProps),
    needRenderPlayer,
    withActiveTab(TABS)
  )(FullMovieCard);

export default ConnectedComponent;
