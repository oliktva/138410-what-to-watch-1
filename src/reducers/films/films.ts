import {Action as ReduxAction} from 'redux';
import {ThunkDispatch as ReduxThunkDispatch, ThunkAction as ReduxThunkAction} from 'redux-thunk';
import {AxiosResponse, AxiosInstance} from 'axios';
import {toCamel} from 'convert-keys';

import {GenreProps} from 'src/types/genres';
import {FilmProps} from 'src/types/films';
import {ReviewProps} from 'src/types/reviews';
import {ThunkDispatch, ThunkAction, State as AppState} from 'src/types/reducer';

export const SET_GENRE = `SET_GENRE`;
export const LOAD_FILMS_SUCCESS = `LOAD_FILMS_SUCCESS`;
export const LOAD_FAVORITE_SUCCESS = `LOAD_FAVORITE_SUCCESS`;
export const UPDATE_FILM = `UPDATE_FILM`;
export const LOAD_PROMO = `LOAD_PROMO`;
export const LOAD_REVIEWS = `LOAD_REVIEWS`;

const ALL_GENRES = `All genres`;

export interface State {
  genre: GenreProps;
  items: FilmProps[];
  favorite: FilmProps[];
  promo: FilmProps | null;
  reviews: {[key: number]: ReviewProps[]} | null;
}

interface SetFilterAction extends ReduxAction {
  type: typeof SET_GENRE;
  payload: GenreProps;
}

interface LoadFilmsAction extends ReduxAction {
  type: typeof LOAD_FILMS_SUCCESS;
  payload: FilmProps[];
}

interface LoadFavoritesAction extends ReduxAction {
  type: typeof LOAD_FAVORITE_SUCCESS;
  payload: FilmProps[];
}

interface UpdateFilmAction extends ReduxAction {
  type: typeof UPDATE_FILM;
  payload: FilmProps;
}

interface LoadFilmPromoAction extends ReduxAction {
  type: typeof LOAD_PROMO;
  payload: FilmProps;
}

interface LoadReviewsAction extends ReduxAction {
  type: typeof LOAD_REVIEWS;
  payload: {id: number; reviews: ReviewProps[]};
}

export type Action =
  SetFilterAction | LoadFilmsAction | LoadFavoritesAction | UpdateFilmAction | LoadFilmPromoAction | LoadReviewsAction;

export type ThunkDispatch = ReduxThunkDispatch<State, AxiosInstance, Action>;
export type ThunkAction = ReduxThunkAction<Promise<void>, State, AxiosInstance, Action>;

export const initialState: State = {
  genre: ALL_GENRES,
  items: [],
  favorite: [],
  promo: null,
  reviews: null
};

const updateFilm = (film: FilmProps, filmsList: FilmProps[]): FilmProps[] =>
  filmsList.map((f: FilmProps): FilmProps => (f.id === film.id ? film : f));

export const ActionCreator = {
  setFilterByGenre: (genre: GenreProps): SetFilterAction => {
    return {
      type: SET_GENRE,
      payload: genre
    };
  },

  loadFilms: (films: FilmProps[]): LoadFilmsAction => {
    return {
      type: LOAD_FILMS_SUCCESS,
      payload: films
    };
  },

  loadFavorite: (films: FilmProps[]): LoadFavoritesAction => {
    return {
      type: LOAD_FAVORITE_SUCCESS,
      payload: films
    };
  },

  updateFilm: (film: FilmProps): UpdateFilmAction => {
    return {
      type: UPDATE_FILM,
      payload: film
    };
  },

  loadPromo: (film: FilmProps): LoadFilmPromoAction => {
    return {
      type: LOAD_PROMO,
      payload: film
    };
  },

  loadReviews: (id: number, reviews: ReviewProps[]): LoadReviewsAction => {
    return {
      type: LOAD_REVIEWS,
      payload: {id, reviews}
    };
  }
};

export const Operation = {
  loadFilms: (): ThunkAction => {
    return (dispatch: ThunkDispatch, _getState: () => AppState, api: AxiosInstance): Promise<void> => {
      return api.get(`/films`)
        .then((response: AxiosResponse<Record<string, any>[]>): void => {
          const data = response.data.map((f): FilmProps => toCamel<FilmProps>(f));

          dispatch(ActionCreator.loadFilms(data));
        });
    };
  },

  loadFavorite: (): ThunkAction => {
    return (dispatch: ThunkDispatch, _getState: () => AppState, api: AxiosInstance): Promise<void> => {
      return api.get(`/favorite`)
        .then((response: AxiosResponse<Record<string, any>[]>): void => {
          const data = response.data.map((f): FilmProps => toCamel<FilmProps>(f));

          dispatch(ActionCreator.loadFavorite(data));
        });
    };
  },

  addToFavorites: (filmId: number): ThunkAction => {
    return (dispatch: ThunkDispatch, _getState: () => AppState, api: AxiosInstance): Promise<void> => {
      return api.post(`/favorite/${filmId}/1`)
        .then((response: AxiosResponse<Record<string, any>>): void => {
          const data = toCamel<FilmProps>(response.data);

          dispatch(ActionCreator.updateFilm(data));
        });
    };
  },

  removeFromFavorites: (filmId: number): ThunkAction => {
    return (dispatch: ThunkDispatch, _getState: () => AppState, api: AxiosInstance): Promise<void> => {
      return api.post(`/favorite/${filmId}/0`)
        .then((response: AxiosResponse<Record<string, any>>): void => {
          const data = toCamel<FilmProps>(response.data);

          dispatch(ActionCreator.updateFilm(data));
        });
    };
  },

  loadPromo: (): ThunkAction => {
    return (dispatch: ThunkDispatch, _getState: () => AppState, api: AxiosInstance): Promise<void> => {
      return api.get(`/films/promo`)
        .then((response: AxiosResponse<Record<string, any>>): void => {
          const data = toCamel<FilmProps>(response.data);

          dispatch(ActionCreator.loadPromo(data));
        });
    };
  },

  loadReviews: (id: number): ThunkAction => {
    return (dispatch: ThunkDispatch, _getState: () => AppState, api: AxiosInstance): Promise<void> => {
      return api.get(`/comments/${id}`)
        .then((response: AxiosResponse<Record<string, any>[]>): void => {
          const data = response.data.map((r): ReviewProps => toCamel<ReviewProps>(r));

          dispatch(ActionCreator.loadReviews(id, data));
        });
    };
  },

  addReview: (id: number, reviewText: string, rating: string | number): ThunkAction => {
    return (dispatch: ThunkDispatch, _getState: () => AppState, api: AxiosInstance): Promise<void> => {
      return api.post(`/comments/${id}`, {
        comment: reviewText,
        rating: typeof rating === `string` ? parseInt(rating, 10) : rating
      }).then((response: AxiosResponse<Record<string, any>[]>): void => {
        const data = response.data.map((r): ReviewProps => toCamel<ReviewProps>(r));

        dispatch(ActionCreator.loadReviews(id, data));
      });
    };
  }
};

const reducer = (state: State = initialState, action: Action): State => {
  switch (action.type) {
    case LOAD_FILMS_SUCCESS:
      return {
        ...state,
        items: action.payload
      };

    case SET_GENRE:
      return {
        ...state,
        genre: action.payload
      };

    case LOAD_FAVORITE_SUCCESS:
      return {
        ...state,
        favorite: action.payload
      };

    case UPDATE_FILM:
      return {
        ...state,
        promo: state.promo ? updateFilm(action.payload, [state.promo])[0] : state.promo,
        items: updateFilm(action.payload, state.items),
        favorite: updateFilm(action.payload, state.favorite)
      };

    case LOAD_PROMO: {
      return {
        ...state,
        promo: action.payload
      };
    }

    case LOAD_REVIEWS: {
      return {
        ...state,
        reviews: {
          ...state.reviews,
          [action.payload.id]: action.payload.reviews
        }
      };
    }

    default:
      return state;
  }
};

export default reducer;
