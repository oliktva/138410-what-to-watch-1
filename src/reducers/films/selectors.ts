import {createSelector} from 'reselect';

import Namespaces from 'src/reducers/namespaces';

import {State} from 'src/types/reducer';
import {FilmProps} from 'src/types/films';
import {GenreProps} from 'src/types/genres';

const ALL_GENRES = `All genres`;

export const getFilms = (state: State): FilmProps[] => {
  return state[Namespaces.FILMS].items;
};

export const getGenre = (state: State): GenreProps => {
  return state[Namespaces.FILMS].genre;
};

export const getFilmsByGenre = createSelector(
  getFilms,
  getGenre,
  (films: FilmProps[], genre: GenreProps): FilmProps[] => {
    return genre === ALL_GENRES ?
      films :
      films.filter((f: FilmProps): boolean => f.genre === genre);
  }
);

export const getGenres = createSelector(
  getFilms,
  (films: FilmProps[]): GenreProps[] => {
    const filmsGenres = films.map((f: FilmProps): GenreProps => f.genre)
      .reduce((result: GenreProps[], genre: GenreProps): GenreProps[] => {
        if (!result.includes(genre)) {
          result.push(genre);
        }

        return result;
      }, []);

    return [ALL_GENRES, ...filmsGenres];
  }
);
