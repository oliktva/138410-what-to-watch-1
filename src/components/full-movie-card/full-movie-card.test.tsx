import React from 'react';
import Enzyme, {shallow} from 'enzyme';
import Adapter from 'enzyme-adapter-react-16';
import toJson from 'enzyme-to-json';

import {FullMovieCard} from './full-movie-card';
import {films, film} from '../../fixtures/films';
import {reviews} from '../../fixtures/reviews';
import {tabs} from '../../fixtures/tabs';
import {history} from '../../fixtures/router';

Enzyme.configure({adapter: new Adapter()});

it(`renders correctly`, () => {
  const fullMovieCard = shallow(<FullMovieCard
    header={<div />}
    film={film}
    relatedFilms={films}
    reviews={reviews}
    withPlayer={false}
    toggleRenderPlayer={jest.fn()}
    addToFavorites={jest.fn()}
    removeFromFavorites={jest.fn()}
    activeTab={tabs[0]}
    setActiveTab={jest.fn()}
    history={history as any}
    match={{} as any}
    location={{} as any}
  />);

  expect(toJson(fullMovieCard)).toMatchSnapshot();
});
