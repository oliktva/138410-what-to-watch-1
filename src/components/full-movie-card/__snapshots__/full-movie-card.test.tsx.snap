// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders correctly 1`] = `
<Fragment>
  <section
    className="movie-card movie-card--full"
    style={
      Object {
        "backgroundColor": "#795433",
      }
    }
  >
    <div
      className="movie-card__hero"
    >
      <BackgroundImage
        image="https://es31-server.appspot.com/wtw/static/film/background/Pulp_Fiction.jpg"
        name="Pulp Fiction"
      />
      <h1
        className="visually-hidden"
      >
        WTW
      </h1>
      <div />
      <div
        className="movie-card__wrap"
      >
        <div
          className="movie-card__desc"
        >
          <MovieMeta
            film={
              Object {
                "backgroundColor": "#795433",
                "backgroundImage": "https://es31-server.appspot.com/wtw/static/film/background/Pulp_Fiction.jpg",
                "description": "The lives of two mob hitmen, a boxer, a gangster & his wife, and a pair of diner bandits intertwine in four tales of violence and redemption.",
                "director": "Quentin Tarantino",
                "genre": "Crime",
                "id": 1,
                "isFavorite": false,
                "name": "Pulp Fiction",
                "posterImage": "https://es31-server.appspot.com/wtw/static/film/poster/Pulp_Fiction.jpg",
                "previewImage": "https://es31-server.appspot.com/wtw/static/film/preview/pulp-fiction.jpg",
                "previewVideoLink": "https://upload.wikimedia.org/wikipedia/commons/transcoded/b/b3/Big_Buck_Bunny_Trailer_400p.ogv/Big_Buck_Bunny_Trailer_400p.ogv.360p.webm",
                "rating": 8.9,
                "released": 1994,
                "runTime": 153,
                "scoresCount": 1635992,
                "starring": Array [
                  "John Travolta",
                  "Uma Thurman",
                  "Samuel L. Jackson",
                ],
                "videoLink": "http://peach.themazzone.com/durian/movies/sintel-1024-surround.mp4",
              }
            }
          />
          <CardButtons
            film={
              Object {
                "backgroundColor": "#795433",
                "backgroundImage": "https://es31-server.appspot.com/wtw/static/film/background/Pulp_Fiction.jpg",
                "description": "The lives of two mob hitmen, a boxer, a gangster & his wife, and a pair of diner bandits intertwine in four tales of violence and redemption.",
                "director": "Quentin Tarantino",
                "genre": "Crime",
                "id": 1,
                "isFavorite": false,
                "name": "Pulp Fiction",
                "posterImage": "https://es31-server.appspot.com/wtw/static/film/poster/Pulp_Fiction.jpg",
                "previewImage": "https://es31-server.appspot.com/wtw/static/film/preview/pulp-fiction.jpg",
                "previewVideoLink": "https://upload.wikimedia.org/wikipedia/commons/transcoded/b/b3/Big_Buck_Bunny_Trailer_400p.ogv/Big_Buck_Bunny_Trailer_400p.ogv.360p.webm",
                "rating": 8.9,
                "released": 1994,
                "runTime": 153,
                "scoresCount": 1635992,
                "starring": Array [
                  "John Travolta",
                  "Uma Thurman",
                  "Samuel L. Jackson",
                ],
                "videoLink": "http://peach.themazzone.com/durian/movies/sintel-1024-surround.mp4",
              }
            }
            isFull={true}
            onFavoritesToggle={[Function]}
            onPlayButtonClick={[Function]}
          />
        </div>
      </div>
    </div>
    <div
      className="movie-card__wrap movie-card__translate-top"
    >
      <div
        className="movie-card__info"
      >
        <MoviePoster
          image="https://es31-server.appspot.com/wtw/static/film/poster/Pulp_Fiction.jpg"
          kind="big"
          name="Pulp Fiction"
        />
        <div
          className="movie-card__desc"
        >
          <nav
            className="movie-nav movie-card__nav"
          >
            <ul
              className="movie-nav__list"
            >
              <li
                className="movie-nav__item "
                key="overview"
              >
                <span
                  className="movie-nav__link"
                  onClick={[Function]}
                >
                  overview
                </span>
              </li>
              <li
                className="movie-nav__item "
                key="details"
              >
                <span
                  className="movie-nav__link"
                  onClick={[Function]}
                >
                  details
                </span>
              </li>
              <li
                className="movie-nav__item "
                key="reviews"
              >
                <span
                  className="movie-nav__link"
                  onClick={[Function]}
                >
                  reviews
                </span>
              </li>
            </ul>
          </nav>
        </div>
      </div>
    </div>
  </section>
  <div
    className="page-content"
  >
    <section
      className="catalog catalog--like-this"
    >
      <h2
        className="catalog__title"
      >
        More like this
      </h2>
      <WithActiveCard
        films={
          Array [
            Object {
              "backgroundColor": "#795433",
              "backgroundImage": "https://es31-server.appspot.com/wtw/static/film/background/Pulp_Fiction.jpg",
              "description": "The lives of two mob hitmen, a boxer, a gangster & his wife, and a pair of diner bandits intertwine in four tales of violence and redemption.",
              "director": "Quentin Tarantino",
              "genre": "Crime",
              "id": 1,
              "isFavorite": false,
              "name": "Pulp Fiction",
              "posterImage": "https://es31-server.appspot.com/wtw/static/film/poster/Pulp_Fiction.jpg",
              "previewImage": "https://es31-server.appspot.com/wtw/static/film/preview/pulp-fiction.jpg",
              "previewVideoLink": "https://upload.wikimedia.org/wikipedia/commons/transcoded/b/b3/Big_Buck_Bunny_Trailer_400p.ogv/Big_Buck_Bunny_Trailer_400p.ogv.360p.webm",
              "rating": 8.9,
              "released": 1994,
              "runTime": 153,
              "scoresCount": 1635992,
              "starring": Array [
                "John Travolta",
                "Uma Thurman",
                "Samuel L. Jackson",
              ],
              "videoLink": "http://peach.themazzone.com/durian/movies/sintel-1024-surround.mp4",
            },
            Object {
              "backgroundColor": "#9B7E61",
              "backgroundImage": "https://es31-server.appspot.com/wtw/static/film/background/War_of_the_Worlds.jpg",
              "description": "As Earth is invaded by alien tripod fighting machines, one family fights for survival.",
              "director": "Steven Spielberg",
              "genre": "Adventure",
              "id": 2,
              "isFavorite": false,
              "name": "War of the Worlds",
              "posterImage": "https://es31-server.appspot.com/wtw/static/film/poster/War_of_the_Worlds.jpg",
              "previewImage": "https://es31-server.appspot.com/wtw/static/film/preview/war-of-the-worlds.jpg",
              "previewVideoLink": "https://upload.wikimedia.org/wikipedia/commons/transcoded/b/b3/Big_Buck_Bunny_Trailer_400p.ogv/Big_Buck_Bunny_Trailer_400p.ogv.360p.webm",
              "rating": 6.5,
              "released": 2005,
              "runTime": 116,
              "scoresCount": 386834,
              "starring": Array [
                "Tom Cruise",
                "Dakota Fanning",
                "Tim Robbins",
              ],
              "videoLink": "http://media.xiph.org/mango/tears_of_steel_1080p.webm",
            },
            Object {
              "backgroundColor": "#A6B7AC",
              "backgroundImage": "https://es31-server.appspot.com/wtw/static/film/background/gangs_of_new_york.jpg",
              "description": "In 1862, Amsterdam Vallon returns to the Five Points area of New York City seeking revenge against Bill the Butcher, his father's killer.",
              "director": "Martin Scorsese",
              "genre": "Crime",
              "id": 3,
              "isFavorite": false,
              "name": "Gangs of new york",
              "posterImage": "https://es31-server.appspot.com/wtw/static/film/poster/Gangs_of_New_York_Poster.jpg",
              "previewImage": "https://es31-server.appspot.com/wtw/static/film/preview/gangs_of_new_york.jpg",
              "previewVideoLink": "https://download.blender.org/durian/trailer/sintel_trailer-480p.mp4",
              "rating": 7.5,
              "released": 2002,
              "runTime": 167,
              "scoresCount": 370881,
              "starring": Array [
                "Leonardo DiCaprio",
                "Cameron Diaz",
                "Daniel Day-Lewis",
              ],
              "videoLink": "http://media.xiph.org/mango/tears_of_steel_1080p.webm",
            },
          ]
        }
      />
    </section>
    <Footer />
  </div>
</Fragment>
`;
